vm/src/object/ext.rs

```diff
@@ -224,6 +225,42 @@ impl<T: PyPayload> ToPyObject for PyRefExact<T> {
     }
 }
 
+pub struct PyAtomicRef<T: PyObjectPayload>(AtomicCell<NonNull<Py<T>>>);
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "threading")] {
+        unsafe impl<T: Send + PyObjectPayload> Send for PyAtomicRef<T> {}
+        unsafe impl<T: Sync + PyObjectPayload> Sync for PyAtomicRef<T> {}
+    }
+}
+
+impl<T: PyObjectPayload> From<PyRef<T>> for PyAtomicRef<T> {
+    fn from(pyref: PyRef<T>) -> Self {
+        let py = PyRef::leak(pyref);
+        Self(AtomicCell::new(NonNull::from(py)))
+    }
+}
+
+impl<T: PyObjectPayload> Deref for PyAtomicRef<T> {
+    type Target = Py<T>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { self.0.load().as_ref() }
+    }
+}
+
+impl<T: PyObjectPayload> PyAtomicRef<T> {
+    /// # Safety
+    /// The caller is responsible to keep the returned PyRef alive
+    /// until no more reference can be used via PyAtomicRef::deref()
+    #[must_use]
+    pub unsafe fn swap(&self, pyref: PyRef<T>) -> PyRef<T> {
+        let py = PyRef::leak(pyref);
+        let old = self.0.swap(NonNull::from(py));
```

Comment:
```suggestion
        let old = self.0.swap(NonNull::new_unchecked(py));
```
we already know py is not a null.

I am not sure possible but retaining `&'static mut T` will be good when available.

Reply:
these two are equal. no check will be made as NonNull::from only accept valid value and it is calling new_unchecked inside the function

Reply:
what do you mean about &'static mut T

Reply:
oh, got it. `py` is a reference, not a pointer.

I was thinking if this is possible.
```
pub struct PyAtomicRef<T: PyObjectPayload>(AtomicCell<&'static mut Py<T>>);
```

Reply:
I think I've found over the past while, mainly from looking at crossbeam's issue list, that atomics aren't actually able to be hidden away enough such that AtomicCell would be a good abstraction. I'd honestly prefer a `PyAtomic<*mut Py<T>>`, which is a type alias that resolves to `std::sync::atomic::AtomicPtr` when threading is enabled, and `Cell<*mut _>` when it's not. I meant to go and replace all the AtomicCell uses with PyAtomic after I added PyAtomic, but I never got around to it.